### ISSUE TITLE
[DISCUSSION] Building HEMCO with CESM/CAM

### ISSUE THREAD

**Issue by jimmielin** [geoschem/HEMCO#20]
## Overview
This is a generic "status update" thread regarding the technical details of an attempt to build HEMCO with the Community Atmosphere Model (CAM).

## Resources
* Haipeng's HEMCO_CESM interface development repository: https://github.com/jimmielin/HEMCO_CESM
* Official ESCOMP HEMCO_CESM repository: https://github.com/ESCOMP/HEMCO_CESM
* CAM: https://github.com/ESCOMP/CAM
* Haipeng's fork of CAM: https://github.com/jimmielin/CAM

Development of a CAM-compatible "fork" of HEMCO (similar to `GEOS` branch) is available in the branch `dev/3.0.0-cesm`. **This branch may be highly unstable and diverging from main code.** The plan is to make changes in this branch, cherry-pick them into `dev/3.0.0` and eventually merge and nuke this testing branch. As such, feel free to play with `dev/3.0.0-cesm` in any way.

## C-Preprocessors
HEMCO in CESM is being compiled with `-DMODEL_ -DMODEL_CESM -DHEMCO_CESM`.
* `MODEL_` is slated to be removed soon. Kept for compatibility with current code.
* `MODEL_CESM` is for compatibility with CESM2-GC.
* `HEMCO_CESM` is a **temporary** flag for HEMCO-CESM code. All code under this is experimental and will be moved to `MODEL_CESM` later.

---

This issue will serve as a posting board so I can update everyone on attempts to compile HEMCO with CAM and the stuff I broke on the way and the status on their eventual fixes. Any feedback & suggestions are welcome at any time!

**Comment by jimmielin** [comment]
I've managed to build HEMCO's source within CAM and link it to the rest of the modules today.

Things I've changed:
* Fixing `regrid_a2a`'s unused IO code (see https://github.com/geoschem/geos-chem/pull/225)
* Commenting out basically the entirety of `HCOIO_Read_Std_Mod` and `HCOIO_Write_Std_Mod` into stubs.
* **Removing** `HCOX_ParaNOx_Mod` (both the module and the bindings in `HCOX_Driver_Mod`)

Plans next: After #11 is finished so HEMCO can accommodate more than 2 IO modules we can start porting in PIO (https://github.com/NCAR/ParallelIO) for use in CAM. The `HCOIO_Read_STD_Mod` code has a bunch of helpers that may need migrating out (#19) in collaboration with @lizziel and @msulprizio before work on this is started, because a lot of the routines are very generic and could be moved elsewhere.

Some work with using PIO in GEOS-Chem's routines (e.g. drydep olson landmap) by Thibaud may be very useful in helping with adding PIO.

**Comment by msulprizio** [comment]
@jimmielin Is this now resolved in HEMCO 3.0.0? Can we now close this discussion?

**Comment by jimmielin** [comment]
Hi @msulprizio the code is in the `dev/cesm` branch. Most changes are incorporated in `13.0.0+final`. We will consolidate the `dev/cesm` changes once GC 13.0.0 is out, so we don't tamper with the code freeze. We can close this now. I will open a new PR for when `dev/cesm` is merged back. Thanks!
